### PR TITLE
Update gradle to new system. #131

### PR DESCRIPTION
**Pull request by lemonboston**
Note that gradle plugin is set to 3.0.1 which probably requires updating Android Studio to that version as well.

### CODE REVIEW

**Comment by lemonboston** [comment]
I've updated this with the dependency config changes as well (squashed and rebased).
Also removed buidToolsVersion from gradle files and from travis config. Let's see if travis can build without that. If not, I will put that back.

**Comment by dmfs** [comment]
shouldn't every module contain this?
```
api project(':contentpal-api')
```
I mean, every module re-exports the contentpal-api. For modules with `api project(':contentpal')` this is transitively "inherited" but `contenttestpal` doesn't have that.

**Comment by lemonboston** [comment]
I am not sure whether we should add `api project(':contentpal-api')` in `contenttestpal` as well. The reason why I didn't do it is because the module that will use `contenttestpal` will have `contentpal` on its main classpath already, so it will be available. Let me know if you think it's better to add it anyway.

This new system is not always obvious for me, for example, `jems` had to exposed in `contentpal`, otherwise the build failed, couldn't find `Optional`. Which kind of makes sense, since those types are used on the interfaces. But the IDE didn't complain, the project compiled there.

I looked up docs again:
https://docs.gradle.org/current/userguide/java_library_plugin.html
> ...distinguish exactly between what is required to compile against the library and what is required to use the library at runtime (in other words, don’t mix what is needed to compile the library itself and what is needed to compile against the library).

I've squashed-rebased, and exposed `jems` in `contentpal-api` as well, even thought the project builds with having that as `implementation` there but I assume that it is needed to build against `contentpal-api` in another project. But I am not certain about that.

File: calendarpal/build.gradle [hunk 1]
```diff
@@ -41,14 +40,16 @@ android {
...
+    implementation 'com.android.support:support-annotations:' + SUPPORT_LIBRARY_VERSION
+    implementation 'org.dmfs:rfc5545-datetime:' + RFC5545_DATETIME_VERSION
+    implementation 'org.dmfs:lib-recur:0.10'
+    implementation 'org.dmfs:jems:' + JEMS_VERSION
+
+    testImplementation project(':contentpal-testing')
+    testImplementation 'junit:junit:' + JUNIT_VERSION
+    testImplementation 'org.mockito:mockito-core:' + MOCKITO_VERSION
+    testImplementation 'org.robolectric:robolectric:' + ROBOLECTRIC_VERSION
+    testImplementation 'org.dmfs:jems-testing:' + JEMS_VERSION
```
Comment (dmfs) [inline calendarpal/build.gradle:54]: this is not necessary when it's already an `implementation` dependency

**Comment by dmfs** [comment]
Ok, other than the minor nit pick I don't have any complains.

I guess `implementation` means these dependencies are no longer transitively passed through. I fully agree with this behavior, if a module uses a dependency it should specify it explicitly as a dependency itself and don't rely on another dependency to pull it in. I think we already discussed this topic about a year ago.

File: calendarpal/build.gradle (hunk 1 above)
Comment (lemonboston) [reply calendarpal/build.gradle:54]: Which one do you mean here? If it was `jems` please note that that's `jems-testing` here.

Comment (dmfs) [reply calendarpal/build.gradle:54]: Oh, right. I missed that. It's ok then.

**Review by dmfs — APPROVED** [review]

**Comment by lemonboston** [comment]
Yep, I absolutely see, too, that this `api` - `implementation` is a clear improvement.
I am thinking though whether it may be needed or will be a good practice to actually separate the interfaces and the implementation of a library to separate modules, always.
Otherwise it may not always be easy to determine what we need to be exposed as `api`. That `jems` in `contentpal` is a good example.
I mean that to determine this:
https://docs.gradle.org/current/userguide/java_library_plugin.html#sec:java_library_separation
>  what is needed to compile the library itself and what is needed to compile against the library

We have to know what exact types from the library are used by the client. I think there is probably a chance that we will run into this a couple of times while doing this migration, that we need to go back to change something from `implementation` to `api` to be able to use a class from the library that uses a type from that dependency.

**Pull request merged**